keps/sig-node/3857-rro-hostpath/README.md

```diff
@@ -0,0 +1,909 @@
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+Support runc's "rro" bind mount option (https://github.com/opencontainers/runc/pull/3272) for Kubernetes's hostPath volumes,
+to make read-only bind mounts literally read-only.
+
+The "rro" bind mount options is implemented by calling [`mount_setattr(2)`](https://man7.org/linux/man-pages/man2/mount_setattr.2.html)
+with `MOUNT_ATTR_RDONLY` and `AT_RECURSIVE`.
+
+Requires runc >= 1.1 && kernel >= 5.12.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+The current `readOnly` hostPath volumes are not recursively read-only;
+e.g., even if `/mnt` is mounted as read-only, its submounts such as `/mnt/usbstorage` are not read-only.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+Support recursively read-only mounts for runc >= 1.1 && kernel >= 5.12.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+Support recursively read-only mounts for old runc and old kernel releases.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A user wants to mount `/mnt`, includings its submounts such as `/mnt/usbstorage`, as read-only.
```

Comment:
Can we expand on the type of workload that may want to do thisv like security scanners?

Reply:
security scanner for what? for malware in the host OS filesystem?

Reply:
Is this KEP rooted in a real use-case, or "this looks neat, we should use it" ?

Do we consider this a security issue?

Reply:
This is a real security issue.

Reply:
I have this lingering fear that maybe we are TOO conservative about making backwards-incompatible changes, especially around security. I really kind of need someone to talk me off the ledge.

Is this issue severe enough that we would risk breaking users with real use cases to get secure-by-default?

Reply:
What about making it enum values like this?

```
readonlyPolicy: (MaybeRecursive|MustBeRecursive|MustNotBeRecursive)
```
- `MaybeRecursive` (default): make it recursively read-only if possible. Otherwise just leave it as-is, without raising an error.
- `MustBeRecursive`: make it recursively read-only, or raise an error. Best for security.
- `MustNotBeRecursive`: Just leave it as-is. Best for compatibility, but probably nobody needs this.

Reply:
We can't make "Maybe" be the default because RRO is a user-visible behavior change, which could break workloads.  The default HAS TO BE the historical behavior, codified.

I agree this sort of tri-state makes sense.

Do we make this specifically about RRO, e.g.

```
recursivelyReadOnly: (Disabled|IfPossible|Required)
```

Or do we try to fold `readOnly` into it, e.g.

```
readOnlyMode: (Writeable | ReadOnlyShallow | ReadOnlyStrongestAvailable | ReadOnlyRecursive)
```

What I am worried about is that we will tell people "you should set the strictest option you can" and then they have to set both `readOnly` AND `readOnlyMode`.

We could change the defaults logic (make `readOnly` a `*bool`) to something like:

```
if readOnly == nil  && readOnlyMode == nil {
    // neither was specified
    readOnly = &roFalse)
    readOnlyMode = &roWriteable
} else if readOnly == nil {
    // readOnlyMode was specified, readOnly was not
    if *readOnlyMode == Writeable {
        readOnly = &roFalse
    } else {
        readOnly = &roTrue
    }
} else if readOnlyMode == nil {
    // readOnly was specified, readOnlyMode was not
    if *readOnly {
        readOnlyMode = &roShallow
    }  else {
        readOnlyMode = &roWriteable
} else {
    // both were specified - validation will ensure they are compatible
}
```
    
I don't love the names, but then you only need to specify one or the other field, never both.

Thoughts?

Reply:
yes.  We TRY not to do that very often, but frankly, we don't have hard rules on this.

Reply:
Was there any such change in the past? I fear it will make a bunch of projects uncompilable.

Reply:
Yes, we have made many such changes.  It sucks but honestly the way we do client lib is totally broken :(

Reply:
Updated the PR to adopt `recursivelyReadOnly: (Disabled|IfPossible|Required)`. Thanks for your suggestion.

`readOnlyMode: (Writeable | ReadOnlyShallow | ReadOnlyStrongestAvailable | ReadOnlyRecursive)` seems a little bit complex to me.

Reply:
Finally catching up with PRs.

Doing this as `readOnly bool` plus `recursivelyReadOnly` means users need to set two linked fields for the same concept.  Doing it the other way changes a client field, but provides a cleaner path.  (incidentally, this is ANOTHER example of why bool fields are bad :)

@msau42 for API opinion - since this is sort of in your wheelhouse anyway :)

Reply:
> Doing this as readOnly bool plus recursivelyReadOnly means users need to set two linked fields for the same concept.

Yes, but just one line, and YAMLs have been already redundant 🙂

Reply:
> We also probably need a way for nodes to indicate that they do support it, so a node status field? Not sure if there is precedent for this. @mrunalp @bobbypage might know.

Can we use `StatusResponse.info`, or do we need to add another property to `StatusResponse`?
https://github.com/kubernetes/cri-api/blob/accd48937811380c1bdc4856a4e6373edb95dbed/pkg/apis/runtime/v1/api.proto#L1491-L1499

Reply:
Thinking of a change like this
```diff
diff --git a/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto b/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
index 5869ebc5263..5f9472168c0 100644
--- a/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
+++ b/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
@@ -212,6 +212,12 @@ enum MountPropagation {
     PROPAGATION_BIDIRECTIONAL = 2;
 }
 
+enum MountRecursivelyReadOnly {
+  RRO_DISABLED = 0;
+  RRO_IF_POSSIBLE = 1;
+  RRO_REQUIRED = 2;
+}
+
 // Mount specifies a host volume to mount into a container.
 message Mount {
     // Path of the mount within the container.
@@ -226,6 +232,7 @@ message Mount {
     bool selinux_relabel = 4;
     // Requested propagation mode.
     MountPropagation propagation = 5;
+    MountRecursivelyReadOnly recursively_read_only = 6;
 }
 
 // IDMapping describes host to container ID mappings for a pod sandbox.
@@ -1491,6 +1498,12 @@ message StatusRequest {
     bool verbose = 1;
 }
 
+// RuntimeHandlerFeatures describes the features that are supported by the runtime handler.
+message RuntimeHandlerFeatures {
+    // RecursivelyReadOnly is true if `runc features | jq .mountOptions` contains "rro".
+    bool RecursivelyReadOnly = 1;
+}
+
 message StatusResponse {
     // Status of the Runtime.
     RuntimeStatus status = 1;
@@ -1499,6 +1512,8 @@ message StatusResponse {
     // debug, e.g. plugins used by the container runtime.
     // It should only be returned non-empty when Verbose is true.
     map<string, string> info = 2;
+    // key = runtime handler name. The empty string denotes the default handler.
+    map<string, RuntimeHandlerFeatures> features = 3;
 }
 
 message ImageFsInfoRequest {}
```

`map<string, RuntimeHandlerFeatures> features` should be probably added to the `nodeInfo` property (and/or labels) of the Node object in the kube-apiserver too

Reply:
Updated the PR to include above

Reply:
> Doing this as `readOnly bool` plus `recursivelyReadOnly` means users need to set two linked fields for the same concept. Doing it the other way changes a client field, but provides a cleaner path. 

I do like the approach of having a new field that also encompasses `readOnly` and then backfilling the `readOnly` field. However, I'm not sure if we can elevate a non-pointer field to a pointer field without breaking a lot of users. Can we backfill a non-pointer field?

Reply:
> Can we backfill a non-pointer field?

Not sure whether it is consistent with the existing design of the yamls.
Is there any existing non-pointer field that is already backfilled?

Reply:
The only real problem with changing a non-pointer to pointer is the client library.  If we don't do that, we can't tell "I didn't set this field" from "I set it to false".  We don't generally change things that a user set explicitly.  This is also why we (now) tend to avoid bool fields.

I think it's fair game, personally but I will discuss more on slack.

Reply:
Changed `ReadOnly bool` to `*bool`.